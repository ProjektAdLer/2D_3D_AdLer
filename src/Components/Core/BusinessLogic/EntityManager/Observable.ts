export default abstract class Observable<T> {
  protected value: T;
  protected subscribers: ((value: T) => void)[] = [];

  public subscribe(fn: (value: T) => void): void {
    this.subscribers.push(fn);
  }

  public unsubscribe(fn: (value: T) => void): void {
    this.subscribers = this.subscribers.filter((subscriber) => {
      return subscriber !== fn;
    });
  }

  protected notify(data: T): void {
    this.subscribers.forEach((fn) => {
      fn(data);
    });
  }

  get Value(): T {
    return this.value;
  }
}
