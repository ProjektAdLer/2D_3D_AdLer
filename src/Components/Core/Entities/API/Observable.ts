export default class Observable<T> {
  constructor(data?: T) {
    if (data) {
      this.data = data;
    }
  }
  private data: T;
  private subscribers: Array<(data: T) => void> = [];

  public subscribe(fn: (data: T) => void): void {
    this.subscribers.push(fn);
  }

  public unsubscribe(fn: (data: T) => void): void {
    this.subscribers = this.subscribers.filter((subscriber) => {
      return subscriber !== fn;
    });
  }

  public notify(data: T): void {
    this.subscribers.forEach((fn) => {
      fn(data);
    });
  }

  public getData(): T {
    return this.data;
  }

  public setData(data: T): void {
    this.data = data;
    this.notify(data);
  }
}
