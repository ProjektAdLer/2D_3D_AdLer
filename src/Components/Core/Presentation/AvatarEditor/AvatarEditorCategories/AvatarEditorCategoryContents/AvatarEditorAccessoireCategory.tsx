import AvatarEditorCategoryContentProps from "./AvatarEditorCategoryContentProps";
import TileGridLayout from "~ReactComponents/GeneralComponents/TileLayout/TileGridLayout";
import { useTranslation } from "react-i18next";
import AccordionElement from "~ReactComponents/GeneralComponents/Accordion/AccordionElement";
import {
  AvatarNoneModel,
  OAvatarHeadGearModels,
  AvatarHeadgearModels,
  OAvatarGlassesModels,
  AvatarGlassesModels,
  OAvatarBackpackModels,
  AvatarBackpackModels,
  OAvatarOtherModels,
  AvatarOtherModels,
} from "src/Components/Core/Domain/AvatarModels/AvatarModelTypes";
import noneIcon from "../../../../../../Assets/avatarEditorThumbnails/aa-none_Thumbnail.png";
import useObservable from "~ReactComponents/ReactRelated/CustomHooks/useObservable";

const noneThumbnail = {
  type: AvatarNoneModel.None,
  image: noneIcon,
};

const headgearThumbnails = Object.values(OAvatarHeadGearModels).map<{
  type: AvatarHeadgearModels;
  image: string;
}>((type) => ({
  type: type,
  image: require(
    `../../../../../../Assets/avatarEditorThumbnails/accessoires/headgear/aa-${type}.png`,
  ),
}));
headgearThumbnails.unshift(noneThumbnail);

const glassesThumbnails = Object.values(OAvatarGlassesModels).map<{
  type: AvatarGlassesModels;
  image: string;
}>((type) => ({
  type: type,
  image: require(
    `../../../../../../Assets/avatarEditorThumbnails/accessoires/glasses/aa-${type}.png`,
  ),
}));
glassesThumbnails.unshift(noneThumbnail);

const backpackThumbnails = Object.values(OAvatarBackpackModels).map<{
  type: AvatarBackpackModels;
  image: string;
}>((type) => ({
  type: type,
  image: require(
    `../../../../../../Assets/avatarEditorThumbnails/accessoires/backpack/aa-${type}.png`,
  ),
}));
backpackThumbnails.unshift(noneThumbnail);

const otherThumbnails = Object.values(OAvatarOtherModels).map<{
  type: AvatarOtherModels;
  image: string;
}>((type) => ({
  type: type,
  image: require(
    `../../../../../../Assets/avatarEditorThumbnails/accessoires/other/aa-${type}.png`,
  ),
}));
otherThumbnails.unshift(noneThumbnail);

export default function AvatarEditorAccessoireCategory(
  props: AvatarEditorCategoryContentProps,
) {
  const { t: translate } = useTranslation("avatarEditor");
  const [headgearType] = useObservable(props.viewModel.headgear);
  const [glassesType] = useObservable(props.viewModel.glasses);
  const [backpackType] = useObservable(props.viewModel.backpack);
  const [otherType] = useObservable(props.viewModel.other);

  const TileGridHeadGear = () => {
    return (
      <TileGridLayout
        tileContents={headgearThumbnails.map((thumbnail, index) => ({
          id: index,
          image: thumbnail.image,
          title: translate(thumbnail.type).toString() ?? "",
          active: headgearType === thumbnail.type,
        }))}
        columns={5}
        mobileColumns={3}
        onTileClick={(id) => {
          props.controller.onAvatarConfigChanged({
            headgear: headgearThumbnails[id].type,
          });
        }}
      />
    );
  };

  const TileGridGlasses = () => {
    return (
      <TileGridLayout
        tileContents={glassesThumbnails.map((thumbnail, index) => ({
          id: index,
          image: thumbnail.image,
          title: translate(thumbnail.type).toString() ?? "",
          active: glassesType === thumbnail.type,
        }))}
        columns={5}
        mobileColumns={3}
        onTileClick={(id) => {
          props.controller.onAvatarConfigChanged({
            glasses: glassesThumbnails[id].type,
          });
        }}
      />
    );
  };

  const TileGridBackpacks = () => {
    return (
      <TileGridLayout
        tileContents={backpackThumbnails.map((thumbnail, index) => ({
          id: index,
          image: thumbnail.image,
          title: translate(thumbnail.type).toString() ?? "",
          active: backpackType === thumbnail.type,
        }))}
        columns={5}
        mobileColumns={3}
        onTileClick={(id) => {
          props.controller.onAvatarConfigChanged({
            backpack: backpackThumbnails[id].type,
          });
        }}
      />
    );
  };

  const TileGridOthers = () => {
    return (
      <TileGridLayout
        tileContents={otherThumbnails.map((thumbnail, index) => ({
          id: index,
          image: thumbnail.image,
          title: translate(thumbnail.type).toString() ?? "",
          active: otherType === thumbnail.type,
        }))}
        columns={5}
        mobileColumns={3}
        onTileClick={(id) => {
          props.controller.onAvatarConfigChanged({
            other: otherThumbnails[id].type,
          });
        }}
      />
    );
  };
  if (!props.controller || !props.viewModel) return null;
  return (
    <div className="flex flex-col">
      <AccordionElement
        header={translate("headgearTitle").toString()}
        isOpen={props.viewModel.uiVisiblity.accessoireMenu.headgear}
        content={TileGridHeadGear()}
        toolTip={translate("accordionCategoryToolTip", {
          subCategory: translate("headgearTitle").toString(),
        }).toString()}
      />
      <AccordionElement
        header={translate("glassesTitle").toString()}
        isOpen={props.viewModel.uiVisiblity.accessoireMenu.glasses}
        content={TileGridGlasses()}
        toolTip={translate("accordionCategoryToolTip", {
          subCategory: translate("glassesTitle").toString(),
        }).toString()}
      />
      <AccordionElement
        header={translate("backpackTitle").toString()}
        isOpen={props.viewModel.uiVisiblity.accessoireMenu.backpack}
        content={TileGridBackpacks()}
        toolTip={translate("accordionCategoryToolTip", {
          subCategory: translate("backpackTitle").toString(),
        }).toString()}
      />
      <AccordionElement
        header={translate("otherTitle").toString()}
        isOpen={props.viewModel.uiVisiblity.accessoireMenu.other}
        content={TileGridOthers()}
        toolTip={translate("accordionCategoryToolTip", {
          subCategory: translate("otherTitle").toString(),
        }).toString()}
      />
    </div>
  );
}
