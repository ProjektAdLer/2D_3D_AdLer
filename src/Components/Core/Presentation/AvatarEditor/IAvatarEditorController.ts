import AvatarConfigTO from "../../Application/DataTransferObjects/AvatarConfigTO";

export default interface IAvatarEditorController {
  onAvatarConfigChanged(changes: Partial<AvatarConfigTO>): void;
  saveAvatarConfig(): void;
  resetAvatarConfig(): void;
}
