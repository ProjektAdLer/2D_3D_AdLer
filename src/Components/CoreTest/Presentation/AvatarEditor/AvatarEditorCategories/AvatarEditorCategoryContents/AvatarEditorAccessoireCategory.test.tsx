import { mock } from "jest-mock-extended";
import AvatarEditorViewModel, {
  AvatarEditorUI,
} from "../../../../../Core/Presentation/AvatarEditor/AvatarEditorViewModel";
import IAvatarEditorController from "../../../../../Core/Presentation/AvatarEditor/IAvatarEditorController";
import Observable from "../../../../../../Lib/Observable";
import {
  AvatarBackpackModels,
  AvatarGlassesModels,
  AvatarHeadgearModels,
  AvatarOtherModels,
} from "../../../../../Core/Domain/AvatarModels/AvatarModelTypes";
import React from "react";
import { fireEvent, render } from "@testing-library/react";
import AvatarEditorAccessoireCategory from "../../../../../Core/Presentation/AvatarEditor/AvatarEditorCategories/AvatarEditorCategoryContents/AvatarEditorAccessoireCategory";

const avatarEditorMock = mock<AvatarEditorViewModel>();
const avatarEditorControllerMock = mock<IAvatarEditorController>();

describe("AvatarEditorHairCategory", () => {
  // ANF-ID: [EZZ0041, EZZ0049, ELG0035, ELG0036]
  test("should render", () => {
    avatarEditorMock.uiVisiblity = {
      accessoireMenu: {
        headgear: new Observable<boolean>(true),
        glasses: new Observable<boolean>(true),
        backpack: new Observable<boolean>(true),
        other: new Observable<boolean>(true),
      },
    } as AvatarEditorUI;

    avatarEditorMock.headgear = new Observable<AvatarHeadgearModels>("none");
    avatarEditorMock.glasses = new Observable<AvatarGlassesModels>("none");
    avatarEditorMock.backpack = new Observable<AvatarBackpackModels>("none");
    avatarEditorMock.other = new Observable<AvatarOtherModels>("none");

    const { container } = render(
      <AvatarEditorAccessoireCategory
        viewModel={avatarEditorMock}
        controller={avatarEditorControllerMock}
      />,
    );
    expect(container).toMatchSnapshot();
  });

  // ANF-ID: [EZZ0041, EZZ0049, ELG0035, ELG0036]
  test.each([
    ["headgear", true, false, false, false],
    ["glasses", false, true, false, false],
    ["backpack", false, false, true, false],
    ["other", false, false, false, true],
  ])(
    "click on tile in category %s calls controller",
    (category, headgearUI, glassesUI, backpackUI, otherUI) => {
      avatarEditorMock.uiVisiblity = {
        accessoireMenu: {
          headgear: new Observable<boolean>(headgearUI),
          glasses: new Observable<boolean>(glassesUI),
          backpack: new Observable<boolean>(backpackUI),
          other: new Observable<boolean>(otherUI),
        },
      } as AvatarEditorUI;

      avatarEditorMock.headgear = new Observable<AvatarHeadgearModels>("none");
      avatarEditorMock.glasses = new Observable<AvatarGlassesModels>("none");
      avatarEditorMock.backpack = new Observable<AvatarBackpackModels>("none");
      avatarEditorMock.other = new Observable<AvatarOtherModels>("none");

      const container = render(
        <AvatarEditorAccessoireCategory
          viewModel={avatarEditorMock}
          controller={avatarEditorControllerMock}
        />,
      );

      const tile = container.getByAltText("noneThumbnail");
      fireEvent.click(tile);

      expect(
        avatarEditorControllerMock.onAvatarConfigChanged,
      ).toHaveBeenCalled();
    },
  );
  test("should not render if controller is undefined", () => {
    const container = render(
      <AvatarEditorAccessoireCategory
        viewModel={avatarEditorMock}
        controller={undefined!}
      />,
    );
    expect(container.container).toBeEmptyDOMElement();
  });
});
