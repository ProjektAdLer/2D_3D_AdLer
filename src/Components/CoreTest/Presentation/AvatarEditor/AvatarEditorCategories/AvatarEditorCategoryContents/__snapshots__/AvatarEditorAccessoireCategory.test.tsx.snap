// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AvatarEditorHairCategory should render 1`] = `
<div>
  <div
    class="flex flex-col"
  >
    <div
      class="pb-2 border-b border-gray-500"
    >
      <button
        class="content-start w-full text-left"
        title="accordionCategoryToolTip"
      >
        <h1
          class="text-2xl portrait:text-lg font-bold"
        >
          ▼ headgearTitle
        </h1>
      </button>
    </div>
    <div>
      <div
        class="accordionOpening"
      >
        <div
          class="grid gap-5 pt-4 rounded-lg p-4 portrait:grid-cols-3 grid-cols-5"
        >
          <div
            class="w-full h-full"
          >
            <button
              class="flex-col px-2 py-1 justify-center flex items-center text-sm rounded-lg hover:cursor-pointer lg:text-xl transition ease-in-out duration-75 active:translate-x-[1px] active:translate-y-[1px] active:border-b-2 active:border-r-2 active:border-transparent text-adlerdarkblue font-regular border-b-2 border-r-2 border-adlerdarkblue overflow-hidden box-border cursor-pointer bg-adleryellow landscape:hover:bg-adleryellow active:bg-adleryellow"
              data-testid="none"
              title="none"
            >
              <div
                class="flex flex-col"
              >
                <img
                  alt="noneThumbnail"
                  class="w-full h-full"
                  src="aa-none_Thumbnail.png"
                />
              </div>
            </button>
          </div>
          <div
            class="w-full h-full"
          >
            <button
              class="flex-col px-2 py-1 justify-center flex items-center text-sm rounded-lg hover:cursor-pointer lg:text-xl transition ease-in-out duration-75 active:translate-x-[1px] active:translate-y-[1px] active:border-b-2 active:border-r-2 active:border-transparent text-adlerdarkblue font-regular border-b-2 border-r-2 border-adlerdarkblue overflow-hidden box-border cursor-pointer bg-buttonbgblue landscape:hover:bg-adleryellow active:bg-adleryellow"
              data-testid="hats-cowboy"
              title="hats-cowboy"
            >
              <div
                class="flex flex-col"
              >
                <img
                  alt="hats-cowboyThumbnail"
                  class="w-full h-full"
                  src="aa-hats-cowboy.png"
                />
              </div>
            </button>
          </div>
          <div
            class="w-full h-full"
          >
            <button
              class="flex-col px-2 py-1 justify-center flex items-center text-sm rounded-lg hover:cursor-pointer lg:text-xl transition ease-in-out duration-75 active:translate-x-[1px] active:translate-y-[1px] active:border-b-2 active:border-r-2 active:border-transparent text-adlerdarkblue font-regular border-b-2 border-r-2 border-adlerdarkblue overflow-hidden box-border cursor-pointer bg-buttonbgblue landscape:hover:bg-adleryellow active:bg-adleryellow"
              data-testid="hats-cylinder"
              title="hats-cylinder"
            >
              <div
                class="flex flex-col"
              >
                <img
                  alt="hats-cylinderThumbnail"
                  class="w-full h-full"
                  src="aa-hats-cylinder.png"
                />
              </div>
            </button>
          </div>
          <div
            class="w-full h-full"
          >
            <button
              class="flex-col px-2 py-1 justify-center flex items-center text-sm rounded-lg hover:cursor-pointer lg:text-xl transition ease-in-out duration-75 active:translate-x-[1px] active:translate-y-[1px] active:border-b-2 active:border-r-2 active:border-transparent text-adlerdarkblue font-regular border-b-2 border-r-2 border-adlerdarkblue overflow-hidden box-border cursor-pointer bg-buttonbgblue landscape:hover:bg-adleryellow active:bg-adleryellow"
              data-testid="hats-froghat"
              title="hats-froghat"
            >
              <div
                class="flex flex-col"
              >
                <img
                  alt="hats-froghatThumbnail"
                  class="w-full h-full"
                  src="aa-hats-froghat.png"
                />
              </div>
            </button>
          </div>
          <div
            class="w-full h-full"
          >
            <button
              class="flex-col px-2 py-1 justify-center flex items-center text-sm rounded-lg hover:cursor-pointer lg:text-xl transition ease-in-out duration-75 active:translate-x-[1px] active:translate-y-[1px] active:border-b-2 active:border-r-2 active:border-transparent text-adlerdarkblue font-regular border-b-2 border-r-2 border-adlerdarkblue overflow-hidden box-border cursor-pointer bg-buttonbgblue landscape:hover:bg-adleryellow active:bg-adleryellow"
              data-testid="hats-kecap"
              title="hats-kecap"
            >
              <div
                class="flex flex-col"
              >
                <img
                  alt="hats-kecapThumbnail"
                  class="w-full h-full"
                  src="aa-hats-kecap.png"
                />
              </div>
            </button>
          </div>
          <div
            class="w-full h-full"
          >
            <button
              class="flex-col px-2 py-1 justify-center flex items-center text-sm rounded-lg hover:cursor-pointer lg:text-xl transition ease-in-out duration-75 active:translate-x-[1px] active:translate-y-[1px] active:border-b-2 active:border-r-2 active:border-transparent text-adlerdarkblue font-regular border-b-2 border-r-2 border-adlerdarkblue overflow-hidden box-border cursor-pointer bg-buttonbgblue landscape:hover:bg-adleryellow active:bg-adleryellow"
              data-testid="hats-santa"
              title="hats-santa"
            >
              <div
                class="flex flex-col"
              >
                <img
                  alt="hats-santaThumbnail"
                  class="w-full h-full"
                  src="aa-hats-santa.png"
                />
              </div>
            </button>
          </div>
          <div
            class="w-full h-full"
          >
            <button
              class="flex-col px-2 py-1 justify-center flex items-center text-sm rounded-lg hover:cursor-pointer lg:text-xl transition ease-in-out duration-75 active:translate-x-[1px] active:translate-y-[1px] active:border-b-2 active:border-r-2 active:border-transparent text-adlerdarkblue font-regular border-b-2 border-r-2 border-adlerdarkblue overflow-hidden box-border cursor-pointer bg-buttonbgblue landscape:hover:bg-adleryellow active:bg-adleryellow"
              data-testid="hats-university"
              title="hats-university"
            >
              <div
                class="flex flex-col"
              >
                <img
                  alt="hats-universityThumbnail"
                  class="w-full h-full"
                  src="aa-hats-university.png"
                />
              </div>
            </button>
          </div>
          <div
            class="w-full h-full"
          >
            <button
              class="flex-col px-2 py-1 justify-center flex items-center text-sm rounded-lg hover:cursor-pointer lg:text-xl transition ease-in-out duration-75 active:translate-x-[1px] active:translate-y-[1px] active:border-b-2 active:border-r-2 active:border-transparent text-adlerdarkblue font-regular border-b-2 border-r-2 border-adlerdarkblue overflow-hidden box-border cursor-pointer bg-buttonbgblue landscape:hover:bg-adleryellow active:bg-adleryellow"
              data-testid="hats-vrgoggle"
              title="hats-vrgoggle"
            >
              <div
                class="flex flex-col"
              >
                <img
                  alt="hats-vrgoggleThumbnail"
                  class="w-full h-full"
                  src="aa-hats-vrgoggle.png"
                />
              </div>
            </button>
          </div>
          <div
            class="w-full h-full"
          >
            <button
              class="flex-col px-2 py-1 justify-center flex items-center text-sm rounded-lg hover:cursor-pointer lg:text-xl transition ease-in-out duration-75 active:translate-x-[1px] active:translate-y-[1px] active:border-b-2 active:border-r-2 active:border-transparent text-adlerdarkblue font-regular border-b-2 border-r-2 border-adlerdarkblue overflow-hidden box-border cursor-pointer bg-buttonbgblue landscape:hover:bg-adleryellow active:bg-adleryellow"
              data-testid="hats-witch"
              title="hats-witch"
            >
              <div
                class="flex flex-col"
              >
                <img
                  alt="hats-witchThumbnail"
                  class="w-full h-full"
                  src="aa-hats-witch.png"
                />
              </div>
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      class="pb-2 border-b border-gray-500"
    >
      <button
        class="content-start w-full text-left"
        title="accordionCategoryToolTip"
      >
        <h1
          class="text-2xl portrait:text-lg font-bold"
        >
          ▼ glassesTitle
        </h1>
      </button>
    </div>
    <div>
      <div
        class="accordionOpening"
      >
        <div
          class="grid gap-5 pt-4 rounded-lg p-4 portrait:grid-cols-3 grid-cols-5"
        >
          <div
            class="w-full h-full"
          >
            <button
              class="flex-col px-2 py-1 justify-center flex items-center text-sm rounded-lg hover:cursor-pointer lg:text-xl transition ease-in-out duration-75 active:translate-x-[1px] active:translate-y-[1px] active:border-b-2 active:border-r-2 active:border-transparent text-adlerdarkblue font-regular border-b-2 border-r-2 border-adlerdarkblue overflow-hidden box-border cursor-pointer bg-adleryellow landscape:hover:bg-adleryellow active:bg-adleryellow"
              data-testid="none"
              title="none"
            >
              <div
                class="flex flex-col"
              >
                <img
                  alt="noneThumbnail"
                  class="w-full h-full"
                  src="aa-none_Thumbnail.png"
                />
              </div>
            </button>
          </div>
          <div
            class="w-full h-full"
          >
            <button
              class="flex-col px-2 py-1 justify-center flex items-center text-sm rounded-lg hover:cursor-pointer lg:text-xl transition ease-in-out duration-75 active:translate-x-[1px] active:translate-y-[1px] active:border-b-2 active:border-r-2 active:border-transparent text-adlerdarkblue font-regular border-b-2 border-r-2 border-adlerdarkblue overflow-hidden box-border cursor-pointer bg-buttonbgblue landscape:hover:bg-adleryellow active:bg-adleryellow"
              data-testid="glasses-browline"
              title="glasses-browline"
            >
              <div
                class="flex flex-col"
              >
                <img
                  alt="glasses-browlineThumbnail"
                  class="w-full h-full"
                  src="aa-glasses-browline.png"
                />
              </div>
            </button>
          </div>
          <div
            class="w-full h-full"
          >
            <button
              class="flex-col px-2 py-1 justify-center flex items-center text-sm rounded-lg hover:cursor-pointer lg:text-xl transition ease-in-out duration-75 active:translate-x-[1px] active:translate-y-[1px] active:border-b-2 active:border-r-2 active:border-transparent text-adlerdarkblue font-regular border-b-2 border-r-2 border-adlerdarkblue overflow-hidden box-border cursor-pointer bg-buttonbgblue landscape:hover:bg-adleryellow active:bg-adleryellow"
              data-testid="glasses-cateye"
              title="glasses-cateye"
            >
              <div
                class="flex flex-col"
              >
                <img
                  alt="glasses-cateyeThumbnail"
                  class="w-full h-full"
                  src="aa-glasses-cateye.png"
                />
              </div>
            </button>
          </div>
          <div
            class="w-full h-full"
          >
            <button
              class="flex-col px-2 py-1 justify-center flex items-center text-sm rounded-lg hover:cursor-pointer lg:text-xl transition ease-in-out duration-75 active:translate-x-[1px] active:translate-y-[1px] active:border-b-2 active:border-r-2 active:border-transparent text-adlerdarkblue font-regular border-b-2 border-r-2 border-adlerdarkblue overflow-hidden box-border cursor-pointer bg-buttonbgblue landscape:hover:bg-adleryellow active:bg-adleryellow"
              data-testid="glasses-dealwithit"
              title="glasses-dealwithit"
            >
              <div
                class="flex flex-col"
              >
                <img
                  alt="glasses-dealwithitThumbnail"
                  class="w-full h-full"
                  src="aa-glasses-dealwithit.png"
                />
              </div>
            </button>
          </div>
          <div
            class="w-full h-full"
          >
            <button
              class="flex-col px-2 py-1 justify-center flex items-center text-sm rounded-lg hover:cursor-pointer lg:text-xl transition ease-in-out duration-75 active:translate-x-[1px] active:translate-y-[1px] active:border-b-2 active:border-r-2 active:border-transparent text-adlerdarkblue font-regular border-b-2 border-r-2 border-adlerdarkblue overflow-hidden box-border cursor-pointer bg-buttonbgblue landscape:hover:bg-adleryellow active:bg-adleryellow"
              data-testid="glasses-heart"
              title="glasses-heart"
            >
              <div
                class="flex flex-col"
              >
                <img
                  alt="glasses-heartThumbnail"
                  class="w-full h-full"
                  src="aa-glasses-heart.png"
                />
              </div>
            </button>
          </div>
          <div
            class="w-full h-full"
          >
            <button
              class="flex-col px-2 py-1 justify-center flex items-center text-sm rounded-lg hover:cursor-pointer lg:text-xl transition ease-in-out duration-75 active:translate-x-[1px] active:translate-y-[1px] active:border-b-2 active:border-r-2 active:border-transparent text-adlerdarkblue font-regular border-b-2 border-r-2 border-adlerdarkblue overflow-hidden box-border cursor-pointer bg-buttonbgblue landscape:hover:bg-adleryellow active:bg-adleryellow"
              data-testid="glasses-monocle"
              title="glasses-monocle"
            >
              <div
                class="flex flex-col"
              >
                <img
                  alt="glasses-monocleThumbnail"
                  class="w-full h-full"
                  src="aa-glasses-monocle.png"
                />
              </div>
            </button>
          </div>
          <div
            class="w-full h-full"
          >
            <button
              class="flex-col px-2 py-1 justify-center flex items-center text-sm rounded-lg hover:cursor-pointer lg:text-xl transition ease-in-out duration-75 active:translate-x-[1px] active:translate-y-[1px] active:border-b-2 active:border-r-2 active:border-transparent text-adlerdarkblue font-regular border-b-2 border-r-2 border-adlerdarkblue overflow-hidden box-border cursor-pointer bg-buttonbgblue landscape:hover:bg-adleryellow active:bg-adleryellow"
              data-testid="glasses-oval"
              title="glasses-oval"
            >
              <div
                class="flex flex-col"
              >
                <img
                  alt="glasses-ovalThumbnail"
                  class="w-full h-full"
                  src="aa-glasses-oval.png"
                />
              </div>
            </button>
          </div>
          <div
            class="w-full h-full"
          >
            <button
              class="flex-col px-2 py-1 justify-center flex items-center text-sm rounded-lg hover:cursor-pointer lg:text-xl transition ease-in-out duration-75 active:translate-x-[1px] active:translate-y-[1px] active:border-b-2 active:border-r-2 active:border-transparent text-adlerdarkblue font-regular border-b-2 border-r-2 border-adlerdarkblue overflow-hidden box-border cursor-pointer bg-buttonbgblue landscape:hover:bg-adleryellow active:bg-adleryellow"
              data-testid="glasses-reading"
              title="glasses-reading"
            >
              <div
                class="flex flex-col"
              >
                <img
                  alt="glasses-readingThumbnail"
                  class="w-full h-full"
                  src="aa-glasses-reading.png"
                />
              </div>
            </button>
          </div>
          <div
            class="w-full h-full"
          >
            <button
              class="flex-col px-2 py-1 justify-center flex items-center text-sm rounded-lg hover:cursor-pointer lg:text-xl transition ease-in-out duration-75 active:translate-x-[1px] active:translate-y-[1px] active:border-b-2 active:border-r-2 active:border-transparent text-adlerdarkblue font-regular border-b-2 border-r-2 border-adlerdarkblue overflow-hidden box-border cursor-pointer bg-buttonbgblue landscape:hover:bg-adleryellow active:bg-adleryellow"
              data-testid="glasses-rectangle"
              title="glasses-rectangle"
            >
              <div
                class="flex flex-col"
              >
                <img
                  alt="glasses-rectangleThumbnail"
                  class="w-full h-full"
                  src="aa-glasses-rectangle.png"
                />
              </div>
            </button>
          </div>
          <div
            class="w-full h-full"
          >
            <button
              class="flex-col px-2 py-1 justify-center flex items-center text-sm rounded-lg hover:cursor-pointer lg:text-xl transition ease-in-out duration-75 active:translate-x-[1px] active:translate-y-[1px] active:border-b-2 active:border-r-2 active:border-transparent text-adlerdarkblue font-regular border-b-2 border-r-2 border-adlerdarkblue overflow-hidden box-border cursor-pointer bg-buttonbgblue landscape:hover:bg-adleryellow active:bg-adleryellow"
              data-testid="glasses-round"
              title="glasses-round"
            >
              <div
                class="flex flex-col"
              >
                <img
                  alt="glasses-roundThumbnail"
                  class="w-full h-full"
                  src="aa-glasses-round.png"
                />
              </div>
            </button>
          </div>
          <div
            class="w-full h-full"
          >
            <button
              class="flex-col px-2 py-1 justify-center flex items-center text-sm rounded-lg hover:cursor-pointer lg:text-xl transition ease-in-out duration-75 active:translate-x-[1px] active:translate-y-[1px] active:border-b-2 active:border-r-2 active:border-transparent text-adlerdarkblue font-regular border-b-2 border-r-2 border-adlerdarkblue overflow-hidden box-border cursor-pointer bg-buttonbgblue landscape:hover:bg-adleryellow active:bg-adleryellow"
              data-testid="glasses-sunglass"
              title="glasses-sunglass"
            >
              <div
                class="flex flex-col"
              >
                <img
                  alt="glasses-sunglassThumbnail"
                  class="w-full h-full"
                  src="aa-glasses-sunglass.png"
                />
              </div>
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      class="pb-2 border-b border-gray-500"
    >
      <button
        class="content-start w-full text-left"
        title="accordionCategoryToolTip"
      >
        <h1
          class="text-2xl portrait:text-lg font-bold"
        >
          ▼ backpackTitle
        </h1>
      </button>
    </div>
    <div>
      <div
        class="accordionOpening"
      >
        <div
          class="grid gap-5 pt-4 rounded-lg p-4 portrait:grid-cols-3 grid-cols-5"
        >
          <div
            class="w-full h-full"
          >
            <button
              class="flex-col px-2 py-1 justify-center flex items-center text-sm rounded-lg hover:cursor-pointer lg:text-xl transition ease-in-out duration-75 active:translate-x-[1px] active:translate-y-[1px] active:border-b-2 active:border-r-2 active:border-transparent text-adlerdarkblue font-regular border-b-2 border-r-2 border-adlerdarkblue overflow-hidden box-border cursor-pointer bg-adleryellow landscape:hover:bg-adleryellow active:bg-adleryellow"
              data-testid="none"
              title="none"
            >
              <div
                class="flex flex-col"
              >
                <img
                  alt="noneThumbnail"
                  class="w-full h-full"
                  src="aa-none_Thumbnail.png"
                />
              </div>
            </button>
          </div>
          <div
            class="w-full h-full"
          >
            <button
              class="flex-col px-2 py-1 justify-center flex items-center text-sm rounded-lg hover:cursor-pointer lg:text-xl transition ease-in-out duration-75 active:translate-x-[1px] active:translate-y-[1px] active:border-b-2 active:border-r-2 active:border-transparent text-adlerdarkblue font-regular border-b-2 border-r-2 border-adlerdarkblue overflow-hidden box-border cursor-pointer bg-buttonbgblue landscape:hover:bg-adleryellow active:bg-adleryellow"
              data-testid="backpack-santapack"
              title="backpack-santapack"
            >
              <div
                class="flex flex-col"
              >
                <img
                  alt="backpack-santapackThumbnail"
                  class="w-full h-full"
                  src="aa-backpack-santapack.png"
                />
              </div>
            </button>
          </div>
          <div
            class="w-full h-full"
          >
            <button
              class="flex-col px-2 py-1 justify-center flex items-center text-sm rounded-lg hover:cursor-pointer lg:text-xl transition ease-in-out duration-75 active:translate-x-[1px] active:translate-y-[1px] active:border-b-2 active:border-r-2 active:border-transparent text-adlerdarkblue font-regular border-b-2 border-r-2 border-adlerdarkblue overflow-hidden box-border cursor-pointer bg-buttonbgblue landscape:hover:bg-adleryellow active:bg-adleryellow"
              data-testid="backpack-threepouchpack"
              title="backpack-threepouchpack"
            >
              <div
                class="flex flex-col"
              >
                <img
                  alt="backpack-threepouchpackThumbnail"
                  class="w-full h-full"
                  src="aa-backpack-threepouchpack.png"
                />
              </div>
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      class="pb-2 border-b border-gray-500"
    >
      <button
        class="content-start w-full text-left"
        title="accordionCategoryToolTip"
      >
        <h1
          class="text-2xl portrait:text-lg font-bold"
        >
          ▼ otherTitle
        </h1>
      </button>
    </div>
    <div>
      <div
        class="accordionOpening"
      >
        <div
          class="grid gap-5 pt-4 rounded-lg p-4 portrait:grid-cols-3 grid-cols-5"
        >
          <div
            class="w-full h-full"
          >
            <button
              class="flex-col px-2 py-1 justify-center flex items-center text-sm rounded-lg hover:cursor-pointer lg:text-xl transition ease-in-out duration-75 active:translate-x-[1px] active:translate-y-[1px] active:border-b-2 active:border-r-2 active:border-transparent text-adlerdarkblue font-regular border-b-2 border-r-2 border-adlerdarkblue overflow-hidden box-border cursor-pointer bg-adleryellow landscape:hover:bg-adleryellow active:bg-adleryellow"
              data-testid="none"
              title="none"
            >
              <div
                class="flex flex-col"
              >
                <img
                  alt="noneThumbnail"
                  class="w-full h-full"
                  src="aa-none_Thumbnail.png"
                />
              </div>
            </button>
          </div>
          <div
            class="w-full h-full"
          >
            <button
              class="flex-col px-2 py-1 justify-center flex items-center text-sm rounded-lg hover:cursor-pointer lg:text-xl transition ease-in-out duration-75 active:translate-x-[1px] active:translate-y-[1px] active:border-b-2 active:border-r-2 active:border-transparent text-adlerdarkblue font-regular border-b-2 border-r-2 border-adlerdarkblue overflow-hidden box-border cursor-pointer bg-buttonbgblue landscape:hover:bg-adleryellow active:bg-adleryellow"
              data-testid="other-sheriff-star"
              title="other-sheriff-star"
            >
              <div
                class="flex flex-col"
              >
                <img
                  alt="other-sheriff-starThumbnail"
                  class="w-full h-full"
                  src="aa-other-sheriff-star.png"
                />
              </div>
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
